toolslm/md_hier.py

```diff
@@ -1,20 +1,37 @@
+"""Parse markdown content into a dictionary with hierarchical keys"""
+
+# AUTOGENERATED! DO NOT EDIT! File to edit: ../04_md_hier.ipynb.
+
+# %% auto 0
+__all__ = ['markdown_to_dict', 'create_heading_dict']
+
+# %% ../04_md_hier.ipynb 2
 import re
 from fastcore.utils import *
-__all__ = ['markdown_to_dict', 'create_heading_dict']
 
-def markdown_to_dict(markdown_content):
+# %% ../04_md_hier.ipynb 3
+def markdown_to_dict(
+    markdown_content: str # The markdown content to parse
+    ) -> dict: # The dictionary with hierarchical keys
+    "Convert a markdown document with headings into a hierarchical dictionary with keys based on the heading structure"
     def clean_heading(text): return re.sub(r'[^A-Za-z0-9 ]+', '', text).strip()
 
     lines = markdown_content.splitlines()
     headings = []
+    in_code_block = False
 
     # Parse headings with their levels and line numbers
     for idx, line in enumerate(lines):
-        match = re.match(r'^(#{1,6})\s*(.*)', line)
-        if match:
-            level = len(match.group(1))
-            text = match.group(2).strip()
-            headings.append({'level': level, 'text': text, 'line': idx})
+        # Toggle code block state when encountering fence
+        if line.strip().startswith('```'): in_code_block = not in_code_block
+        
+        # Only detect headings when not in a code block
+        elif not in_code_block:
+            match = re.match(r'^(#{1,6})\s*(.*)', line)
+            if match:
+                level = len(match.group(1))
+                text = match.group(2).strip()
+                headings.append({'level': level, 'text': text, 'line': idx})
```

Comment:
```suggestion
        if line.strip().startswith('```'): in_code_block = not in_code_block
        
        # Only detect headings when not in a code block
        if in_code_block: continue
        match = re.match(r'^(#{1,6})\s*(.*)', line)
        if match:
            level = len(match.group(1))
            text = match.group(2).strip()
            headings.append({'level': level, 'text': text, 'line': idx})
```

A minor tweak to reduce the indentation level.